File: Python/sql_queries.py
import psycopg2

from crud_operations import CrudOperations


def get_all_books(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT * FROM """ + table_name
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_books_by_author(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str, author: str) -> None:
    query = """
    SELECT * FROM """ + table_name + """ 
    WHERE author = '""" + author + """'"""
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_books_before_year(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str, year: int) -> None:
    query = """
    SELECT * FROM """ + table_name + """ 
    WHERE publication_year < """ + str(year)
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def count_books(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT COUNT(*) FROM """ + table_name
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_books_with_string_in_title(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str, str_search: str) -> None:
    query = """
    SELECT * FROM """ + table_name + """ 
    WHERE title LIKE '%""" + str_search + """%'"""
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_distinct_authors(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT DISTINCT author FROM """ + table_name
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def update_author(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str, old_author: str, new_author: str) -> None:
    query = """
    UPDATE """ + table_name + """ 
    SET author = '""" + new_author + """' 
    WHERE author = '""" + old_author.strip() + """'"""
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def delete_books_before_year(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str, year: int) -> None:
    query = """
    DELETE FROM """ + table_name + """ 
    WHERE publication_year < """ + str(year)
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_all_books_sorted_by_publication_year(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT * FROM """ + table_name + """ 
    ORDER BY publication_year DESC"""
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_oldest_book(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT * FROM """ + table_name + """ 
    ORDER BY publication_year LIMIT 1"""
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def ge_books_by_year(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT publication_year, COUNT(*) 
    FROM """ + table_name + """
    GROUP BY publication_year
    """
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_top_5_authors(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT author, COUNT(*) 
    FROM """ + table_name + """
    GROUP BY author
    ORDER BY COUNT(*) DESC
    LIMIT 5
    """
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_longer_titles(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT title
    FROM """ + table_name + """
    WHERE LENGTH(title) > 15
    """
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_average_publication_year(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT AVG(publication_year)
    FROM """ + table_name + """
    """
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def get_authors_with_multiple_books(con: psycopg2.extensions.connection, table_name: str, queryStr: str, file_path: str) -> None:
    query = """
    SELECT author
    FROM """ + table_name + """
    GROUP BY author, publication_year
    HAVING COUNT(*) > 1
    """
    CrudOperations.execute_and_print_query(con, query, queryStr, file_path)


def run():
    dbName = "bookslibrary"
    user = "postgres"
    password = "admin"
    con = CrudOperations.connect_to_db(dbName, user, password)

    table_name = "books"

    queryStr = None
    file_path = "output.txt"

    # Retrieve all books
    queryStr = "Retrieve all books"
    get_all_books(con, table_name, queryStr, file_path)

    # Find books by a specific author
    queryStr = "Find books by a specific author"
    get_books_by_author(con, table_name, queryStr, file_path, "Onofredo Flay")

    # List all books published before a certain year
    queryStr = "List all books published before a certain year"
    get_books_before_year(con, table_name, queryStr, file_path, 2010)

    # Count the total number of books in the library
    queryStr = "Count the total number of books in the library"
    count_books(con, table_name, queryStr, file_path)

    # Find books with titles containing a specific word
    queryStr = "Find books with titles containing a specific word"
    get_books_with_string_in_title(con, table_name, queryStr, file_path, "sql")

    # List all authors (without duplicates)
    queryStr = "List all authors (without duplicates)"
    get_distinct_authors(con, table_name, queryStr, file_path)

    # Update the author name for all books by a specific author
    queryStr = "Update the author name for all books by a specific author"
    update_author(con, table_name, queryStr, file_path,
                  "Onofredo Flay", "Homer Simpson")

    # Delete all books published before a specific year
    queryStr = "Delete all books published before a specific year"
    delete_books_before_year(con, table_name, queryStr, file_path, 2006)

    # Retrieve all books sorted by publication year (newest first)
    queryStr = "Retrieve all books sorted by publication year (newest first)"
    get_all_books_sorted_by_publication_year(
        con, table_name, queryStr, file_path)

    # Find the oldest book in the library
    queryStr = "Find the oldest book in the library"
    get_oldest_book(con, table_name, queryStr, file_path)

    # Find the number of books published each year
    queryStr = "Find the number of books published each year"
    ge_books_by_year(con, table_name, queryStr, file_path)

    # List the top 5 most frequently occurring authors in the library
    queryStr = "List the top 5 most frequently occurring authors in the library"
    get_top_5_authors(con, table_name, queryStr, file_path)

    # Identify books that have titles longer than 15 characters
    queryStr = "Identify books that have titles longer than 15 characters"
    get_longer_titles(con, table_name, queryStr, file_path)

    # Calculate the average publication year of all books in the library
    queryStr = "Calculate the average publication year of all books in the library"
    get_average_publication_year(con, table_name, queryStr, file_path)

    # Select all authors who have published more than one book in the same year
    queryStr = "Select all authors who have published more than one book in the same year"
    get_authors_with_multiple_books(con, table_name, queryStr, file_path)

    CrudOperations.close_connection(con)


if __name__ == '__main__':
    run()

File: Go/services/sql_queries.go
package services

import (
	"fmt"
	"strings"

	"github.com/jackc/pgx/v5"
	"github.com/siddhant-vij/Personal-Book-Library/utils"
)

func SqlQueries() {
	user := "postgres"
	password := "admin"
	host := "localhost"
	port := "5432"
	dbName := "bookslibrary"

	// Connect to database
	connection, err := utils.ConnectToDB(dbName, user, password, host, port)
	if err != nil {
		fmt.Println(err)
	}

	tableName := "books"
	queryStr := ""
	outputFilePath := "output.txt"

	queryStr = "Retrieve all books"
	getAllBooks(connection, tableName, queryStr, outputFilePath)

	queryStr = "Find books by a specific author"
	getBooksByAuthor(connection, tableName, queryStr, outputFilePath, "Onofredo Flay")

	queryStr = "List all books published before a certain year"
	getBooksBeforeYear(connection, tableName, queryStr, outputFilePath, 2010)

	queryStr = "Count the total number of books in the library"
	countBooks(connection, tableName, queryStr, outputFilePath)

	queryStr = "Find books with titles containing a specific word"
	getBooksWithStringInTitle(connection, tableName, queryStr, outputFilePath, "sql")

	queryStr = "List all authors (without duplicates)"
	getDistinctAuthors(connection, tableName, queryStr, outputFilePath)

	queryStr = "Update the author name for all books by a specific author"
	updateAuthor(connection, tableName, queryStr, outputFilePath, "Onofredo Flay", "Homer Simpson")

	queryStr = "Delete all books published before a specific year"
	deleteBooksBeforeYear(connection, tableName, queryStr, outputFilePath, 2006)

	queryStr = "Retrieve all books sorted by publication year (newest first)"
	getAllBooksSortedByPublicationYear(connection, tableName, queryStr, outputFilePath)

	queryStr = "Find the oldest book in the library"
	getOldestBook(connection, tableName, queryStr, outputFilePath)

	queryStr = "Find the number of books published each year"
	getBooksByYear(connection, tableName, queryStr, outputFilePath)

	queryStr = "List the top 5 most frequently occurring authors in the library"
	getTop5Authors(connection, tableName, queryStr, outputFilePath)

	queryStr = "Identify books that have titles longer than 15 characters"
	getLongerTitles(connection, tableName, queryStr, outputFilePath)

	queryStr = "Calculate the average publication year of all books in the library"
	getAveragePublicationYear(connection, tableName, queryStr, outputFilePath)

	queryStr = "Select all authors who have published more than one book in the same year"
	getAuthorsWithMultipleBooks(connection, tableName, queryStr, outputFilePath)

	// Close database connection
	utils.CloseDB(connection)
}

func performAction(connection *pgx.Conn, queryStr string, query string, outputFilePath string) {
	query = strings.TrimSpace(query)
	if strings.HasPrefix(query, "SELECT") {
		rows, err := utils.ExecuteQuery(connection, query)
		if err != nil {
			fmt.Println(err)
			return
		}
		err = utils.PrintQueryResultToFile(queryStr, rows, outputFilePath)
		if err != nil {
			fmt.Println(err)
			return
		}
	} else {
		count, err := utils.ExecuteUpdateQuery(connection, query)
		if err != nil {
			fmt.Println(err)
			return
		}
		err = utils.PrintRowsAffectedToFile(queryStr, count, outputFilePath)
		if err != nil {
			fmt.Println(err)
			return
		}
	}
}

func getAllBooks(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT * FROM %s", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getBooksByAuthor(con *pgx.Conn, tableName, queryStr, outputFilePath, author string) {
	query := fmt.Sprintf("SELECT * FROM %s WHERE author = '%s'", tableName, author)
	performAction(con, queryStr, query, outputFilePath)
}

func getBooksBeforeYear(con *pgx.Conn, tableName, queryStr, outputFilePath string, year int) {
	query := fmt.Sprintf("SELECT * FROM %s WHERE publication_year < %d", tableName, year)
	performAction(con, queryStr, query, outputFilePath)
}

func countBooks(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT COUNT(*) FROM %s", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getBooksWithStringInTitle(con *pgx.Conn, tableName, queryStr, outputFilePath, title string) {
	query := fmt.Sprintf("SELECT * FROM %s WHERE title LIKE '%%%s%%'", tableName, title)
	performAction(con, queryStr, query, outputFilePath)
}

func getDistinctAuthors(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT DISTINCT author FROM %s", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func updateAuthor(con *pgx.Conn, tableName, queryStr, outputFilePath, oldAuthor, newAuthor string) {
	query := fmt.Sprintf("UPDATE %s SET author = '%s' WHERE author = '%s'", tableName, newAuthor, oldAuthor)
	performAction(con, queryStr, query, outputFilePath)
}

func deleteBooksBeforeYear(con *pgx.Conn, tableName, queryStr, outputFilePath string, year int) {
	query := fmt.Sprintf("DELETE FROM %s WHERE publication_year < %d", tableName, year)
	performAction(con, queryStr, query, outputFilePath)
}

func getAllBooksSortedByPublicationYear(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT * FROM %s ORDER BY publication_year DESC", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getOldestBook(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT * FROM %s ORDER BY publication_year ASC LIMIT 1", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getBooksByYear(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT publication_year, COUNT(*) FROM %s GROUP BY publication_year", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getTop5Authors(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT author, COUNT(*) FROM %s GROUP BY author ORDER BY COUNT(*) DESC LIMIT 5", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getLongerTitles(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT * FROM %s WHERE LENGTH(title) > 15", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getAveragePublicationYear(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT AVG(publication_year) FROM %s", tableName)
	performAction(con, queryStr, query, outputFilePath)
}

func getAuthorsWithMultipleBooks(con *pgx.Conn, tableName, queryStr, outputFilePath string) {
	query := fmt.Sprintf("SELECT author FROM %s GROUP BY author, publication_year HAVING COUNT(*) > 1", tableName)
	performAction(con, queryStr, query, outputFilePath)
}
